### ISSUE TITLE
Add Inline Tooltips

### ISSUE THREAD

**Issue by benniledl** [WordPress/gutenberg#67483]
I have created a plugin ([Inline Tooltips](https://wordpress.org/plugins/inline-tooltips/)) to add tooltips to my Gutenberg sites, and I would love to see such a feature integrated into WordPress Core in the future.

## Tooltip Positioning Options  
It might be beneficial for users to have the option to choose where the tooltip appears relative to the text (top, right, bottom, left). This could solve layout issues, such as when a tooltip is added in a top bar and popping up above the text might cause it to overflow or be cut off. I’m unsure if there are CSS-only approaches to dynamically handle this positioning, but offering this flexibility could make tooltips more reliable in various contexts if there is no better (automatic positioning) option found.  

## Tooltip Styling Options  
The standard `title=""` attribute of HTML tags cannot be styled individually, which is why I opted for a `data-tooltip=""` attribute in my plugin. This allows the tooltip popup to be styled freely with custom CSS. However, I’m not sure if this approach is optimal for SEO purposes. It might be worth exploring whether alternative methods could offer better compatibility with search engines while still allowing for custom styling.  

![Image](https://github.com/user-attachments/assets/1770150f-ce53-41c4-a12d-ccea833bed25)
![Image](https://github.com/user-attachments/assets/47c5b324-a896-400e-85cd-41af87c186a7)

**Comment by TheJeffr0** [comment]
I too would like to see Inline Tooltips as a core feature. Here's an example of how I envision something like this working. https://designsystem.digital.gov/components/tooltip/

**Comment by joedolson** [comment]
If we're going to add something like this, we should probably make use of the `popover` attributes. This might be a good thing to work on for 6.9, as support for `popover` is developing pretty rapidly. But adding a custom integration wouldn't make a lot of sense for the long term.

**Comment by westonruter** [comment]
Note that the Popover API is now [Baseline Newly available](https://web.dev/blog/popover-baseline), although `popover=hint` is currently only supported by Chromium ([caniuse](https://caniuse.com/mdn-api_htmlelement_popover_hint)). The other missing piece for tooltips is **Interest Invokers** ([Chrome Platform Status](https://chromestatus.com/feature/4530756656562176), [explainer](https://open-ui.org/components/interest-invokers.explainer/)).

Something else to consider: sometimes a tooltip should be implemented as `title` while other times as using `popover`. In particular, consider the [`ABBR`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/abbr) element where the `title` attribute has [specific semantic meaning](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/abbr#attributes) what the abbreviation is for. So should there be a separate "Abbreviation" inline to ensure it is used appropriately instead of a more general `popover`? In addition to the `ABBR`, there is also the [`DFN`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dfn) tag for which the `title` attribute also has [special meaning](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dfn#specifying_the_term_being_defined). I'm unsure if an `ABBR` tag with the abbreviation contained in a `popover` would be semantically equivalent to the attribute contained in a `title` attribute, although it seems it can be equivalent from an a11y perspective by using the `aria-labelledby` attribute.

A tooltip implemented with `popover` would allow rich text to be used. In contrast, a `title` attribute only allows plain text. So in the UI if no rich text is entered, would it default to `title` and then only use `popover` if rich text is entered? Would the UI have a toggle for whether it is an abbreviation (per above) in which case rich text wouldn't be allowed?

As for positioning the tooltip, this is specifically what the [Anchor Positioning API](https://developer.chrome.com/blog/anchor-positioning-api) is all about. It allows the popover to be positioned dynamically based on the available viewport area around the target element. It is supported in Chromium and in Safari TP but it is not yet implemented in Firefox ([caniuse](https://caniuse.com/css-anchor-positioning)), _but_ it is [part of Interop 2025](https://web.dev/blog/interop-2025#anchor_positioning) so it should be supported by all browsers this year!

With the Popover API and Anchor Positioning APIs available, this feature could be implemented on the frontend without any JavaScript at all, while also being semantically well-defined and accessible (assuming assistive technologies implement support for them).

As for the UI, I think we could potentially avoid a modal in favor of implementing the UI in the editor itself as a popover. This would allow for a WYSIWYG editing experience. Any toggles for positioning and whether it is for an abbreviation could be located in the block sidebar panel.

**Comment by westonruter** [comment]
In addition to the `ABBR`, there is also the [`DFN`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dfn) tag for which the `title` attribute also has [special meaning](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dfn#specifying_the_term_being_defined).

**Comment by westonruter** [comment]
Related:

* https://github.com/WordPress/gutenberg/issues/54089

**Comment by westonruter** [comment]
In regards to `popover=hint` being the missing piece, it's actually half of the missing piece. The other half is [interest invokers](https://chromestatus.com/feature/4530756656562176). See [explainer](https://open-ui.org/components/interest-invokers.explainer/).

**Comment by westonruter** [comment]
I put together a demo showing how a rich tooltip could look for an `ABBR` using Popover API, Anchor Positioning API, and a polyfill for Interest Invokers: https://codepen.io/westonruter/full/dPyKZdv

**Comment by westonruter** [comment]
Experiment which implements footnotes as popovers: https://github.com/WordPress/gutenberg/pull/69671

**Comment by joedolson** [comment]
From an accessibility perspective, `title` has no special meaning in either the `dfn` or `abbr` elements; the semantics are defined, but not passed through in any unique way to assistive technology. Here's a *very* [detailed exploration of exactly how it's defined and used by Adrian Roselli](https://adrianroselli.com/2024/01/using-abbr-element-with-title-attribute.html).

**Comment by westonruter** [comment]
@joedolson Thanks for that. So consider this generic use of `abbr`:

```html
<p>The joke made me <abbr title="Laugh Out Loud">LOL</abbr> big time.</p>
```

In the Chrome DevTools Accessibility inspector I see the following computed properties for the `abbr` element:

> Name: "Laugh Out Loud"
> * `aria-labelledby`: Not specified
> * `aria-label`: Not specified
> * `title`: "Laugh Out Loud"
>
> Role: Abbr

On the other hand, for the following markup with `popover` (from the [Codepen](https://codepen.io/westonruter/full/dPyKZdv)), which I've adapted further to make use of `aria-describedby`:

```html
<p>The joke made me <abbr style="anchor-name: --lol" aria-labelledby="lol-label" aria-describedby="lol-description"><button popovertarget="lol" interesttarget="lol">LOL</button></abbr> big time.</p>

<aside id="lol" popover="hint" style="position-anchor: --lol; top: anchor(bottom); margin:0; justify-self: anchor-center;">
  <h3 id="lol-label">Laugh Out Loud</h3>
  <div id="lol-description">
    <blockquote cite="https://en.wikipedia.org/wiki/LOL">
      <p><b>LOL</b>, or <b>lol</b>, is an initialism for <b>laughing out loud</b>, and a popular element of <a href="https://en.wikipedia.org/wiki/wiki/Internet_slang" title="Internet slang" target="_blank">Internet slang</a>, which can be used to indicate amusement, irony, or double meanings. <a href="https://en.wikipedia.org/wiki/LOL" target="_blank">Read more</a></p>
    </blockquote>
    <img src="https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExNm51cnVob2huZjZnaGM3czQ4b3NrMHhwZGZqN3I0ZWUwZjNmYngzNCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/XBCJIv6xAyDfrajXoe/giphy.webp" width="240" height="199" alt="Michael Scott from The Office laughing and then holding up his hand to stop" loading="lazy">
  </div>
</aside>
```

The computed properties for this `abbr` element are:

> Name: "Laugh Out Loud"
> * `aria-labelledby`: `h3#lol-label` "Laugh Out Loud"
> * `aria-label`: Not specified
> * `title`: Not specified
>
> Description: "LOL, or lol, is an initialism for laughing out loud, and a popular element of Internet sla..."
> Role: Abbr
> Described by: `div#lol-description`
> details: `aside#lol`
> Labeled by: `h3#lol-label`

In both cases the **name** is "Laugh Out Loud", but in the second case there is also a **description** that has the full definition.

So it seems at least in Chrome these are considered equivalent, or rather that the latter has everything from the former with yet additional information?

**Comment by joedolson** [comment]
The `title` is included in the accessible description calculation, but it's only read optionally by the majority of screen readers. This is largely to do with historic abuse of the attribute. Heuristics have improved over time, but it's still an optional attribute for reading by most screen readers.

Those are also not equivalent because in the first example, the expansion is in a description field, which are added to the accessible name; in the popover example, the use of `aria-labelledby` instead *replaces* the accessible name. I didn't test screen readers to see what they actually do, but in both cases it would override the abbreviation itself. I wouldn't use `aria-labelledby` here. (It doesn't in this case, because hidden elements are ignored when reading.)

In either case, the ARIA attributes should be attached to the `button`, not to the `abbr`. (or any other wrapping semantic element) With the button, they'll be announced when the button is focused; without it, they won't be announced at all, as the `abbr` is treated as an inline style element and not announced at all when reading. Many inline roles are [explicitly prohibited from being named](https://w3c.github.io/aria/#namefromprohibited), and most elements with text-level semantics have either [no or limited support for naming](https://html5accessibility.com/stuff/2022/04/03/aria-labelledby-usage-notes/). 

It's important to be aware that the reading flow for screen reader users will be very different adding a button into the flow. The `abbr` element will not be announced at all by default, so the sentence will be "The joke made me LOL big time." With the button, the sentence as read by NVDA is "The joke made me [button  collapsed] LOL big time. " 

There's a reason that inline elements are mostly not read, and it's because they can significantly interrupt the reading flow. The plus side is that they let you know there's additional information to be found.  You could then choose to focus on the button and get the additional information.

**Comment by westonruter** [comment]
@joedolson Thank you! It seems then that using an `abbr` with a `button` and a `popovertarget` is going to be wrong then. In my [Codepen](https://codepen.io/westonruter/full/dPyKZdv), I had combined the `popovertarget` and the `interesttarget` in the same element, but I've just updated it to split them apart into two separate examples. This is the Interest Invoker example now:

```html
<p>The joke made me <abbr style="anchor-name: --lol2" interesttarget="lol2">LOL</abbr> big time.</p>

<aside id="lol2" popover="hint" style="position-anchor: --lol2; top: anchor(bottom); margin:0; justify-self: anchor-center;">
    <h3>Laugh Out Loud</h3>
    <blockquote cite="https://en.wikipedia.org/wiki/LOL">
        <p>
            <b>LOL</b>, or <b>lol</b>, is an initialism for <b>laughing out loud</b>, and a popular element of <a href="https://en.wikipedia.org/wiki/wiki/Internet_slang" title="Internet slang" target="_blank">Internet slang</a>, which can be used to indicate amusement, irony, or double meanings. <a href="https://en.wikipedia.org/wiki/LOL" target="_blank">Read more</a>
        </p>
    </blockquote>
</aside>
```

It seems like [Interest Invokers](https://open-ui.org/components/interest-invokers.explainer/) should specifically address the problems you've identified. Note the [Accessibility section](https://open-ui.org/components/interest-invokers.explainer/#accessibility). As I understand, it is saying that `interesttarget` should be automatically designated as the `describedby` so that no explicit `aria-describedby="lol2"` is needed on the `abbr`.

Obviously, Interest Invokers are not available in browsers yet, but I'm hoping this can be shown to be a use case for where they could eventually be implemented. In particular, that they could be used to implement rich tooltips in the block editor which are also fully accessible.

**Comment by TheJeffr0** [comment]
Has there been any more work or improvements made since March?
